bpfd/src/oci_utils/cosign.rs

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: (MIT OR Apache-2.0)
+// Copyright Authors of bpfd
+
+use std::str::FromStr;
+
+use anyhow::{anyhow, bail};
+use log::{debug, info, warn};
+use sigstore::{
+    cosign::{
+        verification_constraint::VerificationConstraintVec, verify_constraints, ClientBuilder,
+        CosignCapabilities,
+    },
+    errors::SigstoreError::RegistryPullManifestError,
+    registry::{Auth, ClientConfig, ClientProtocol, OciReference},
+    tuf::SigstoreRepository,
+};
+use tokio::task::spawn_blocking;
+
+pub struct CosignVerifier {
+    pub client: sigstore::cosign::Client,
+    pub allow_unsigned: bool,
+}
+
+impl CosignVerifier {
+    pub(crate) async fn new(allow_unsigned: bool) -> Result<Self, anyhow::Error> {
+        let repo: sigstore::errors::Result<SigstoreRepository> = spawn_blocking(|| {
+            info!("Downloading data from Sigstore TUF repository");
+            sigstore::tuf::SigstoreRepository::fetch(None)
+        })
+        .await
+        .map_err(|e| anyhow!("Error spawning blocking task inside of tokio: {}", e))?;
+
+        let repo: SigstoreRepository = repo?;
+
+        let oci_config = ClientConfig {
+            protocol: ClientProtocol::Https,
+            ..Default::default()
+        };
+
+        let cosign_client = ClientBuilder::default()
+            .with_oci_client_config(oci_config)
+            .with_rekor_pub_key(repo.rekor_pub_key())
+            .with_fulcio_certs(repo.fulcio_certs())
+            .enable_registry_caching()
+            .build()?;
+
+        Ok(Self {
+            client: cosign_client,
+            allow_unsigned,
+        })
+    }
+
+    pub(crate) async fn verify(
+        &mut self,
+        image: &str,
+        username: Option<&str>,
+        password: Option<&str>,
+    ) -> Result<(), anyhow::Error> {
+        debug!("CosignVerifier::verify()");
+        let image = OciReference::from_str(image)?;
+        let auth = if let (Some(username), Some(password)) = (username, password) {
+            Auth::Basic(username.to_string(), password.to_string())
+        } else {
+            Auth::Anonymous
+        };
+
+        debug!("Triangulating image: {}", image);
+        let (cosign_signature_image, source_image_digest) =
+            self.client.triangulate(&image, &auth).await?;
+
+        debug!("Getting trusted layers");
+        match self
+            .client
+            .trusted_signature_layers(&auth, &source_image_digest, &cosign_signature_image)
+            .await
+        {
+            Ok(trusted_layers) => {
+                debug!("Found trusted layers");
+                debug!("Verifying constraints");
+                // TODO: Add some constraints here
```

Comment:
So currently we're not actually verifying anything?

Reply:
We verify whether a signature exists, or not.
We're not yet verifying:
- Whose signature
- Which OIDC provider it came from
- Anything else about the image

All of that will need to come from config somewhere